doc/barclamp_development_exercises.md

```diff
@@ -0,0 +1,295 @@
+
+Crowbar leverages the ``oslo.config`` library to support reading configuration
+from directories under ``/etc/<service>/<service>.conf.d``. The service package
+will install the default config file in ``/etc/<service>/<service>.conf`` and a
+small set of overrides in
+``/etc/<service>/<service>.conf.d/010-<service>.conf``. Crowbar then applies its
+service configuration in ``/etc/<service>/<service>.conf.d/100-<service>.conf``
+which overrides all configuration in the default config file or the package's
+override config file. Because Chef is an traditional-style configuration
+management tool, it attempts to converge a server's configuration by reapplying
+recipes regularly, about every 15 minutes or so. This means that any local
+changes you make to the ``100-<service>.conf`` files that Crowbar manages will
+be eventually overwritten. Using override config files is a useful way of
+working around this, since those files will not be overwritten. Override config
+files are also useful for handling any parameters not yet supported by Crowbar,
+or for temporarily applying a setting without going through a barclamp proposal.
+
+### Exercise
+
+Create a file ``/etc/keystone/keystone.conf.d/101-debug.conf``.
+
+In the file, add the content:
+
+```
+[DEFAULT]
+debug = true
+insecure_debug = true
+```
+
+Restart the keystone service:
+
+```
+# systemctl restart apache2
+```
+
+The ``insecure_debug`` option is now enabled in keystone, which Crowbar doesn't
+support. This option is useful for debugging authentication requests on
+non-production systems. You can see it in action by requesting tokens from
+keystone using invalidate credentials or incorrect scopes.
+
+## Adding a Parameter to a Barclamp
+
+Adding a new parameter involves making changes to several components, but is
+ultimately a matter of copying and pasting similar examples. This exercise will
+add the ``insecure_debug`` parameter to the keystone barclamp. In reality we
+should never support setting this option through Crowbar because it is insecure
+for production use, but it is useful as an exercise.
+
+### Exercise
+
+This is done in three stages.
+
+#### Add the parameter to the data bag
+
+Edit the keystone data bag schema in
+``/opt/dell/chef/data_bags/crowbar/template-keystone.schema``. Add a new
+parameter ``insecure_debug`` under the existing ``debug`` parameter.
+
+Edit the default data bag in
+``/opt/dell/chef/data_bags/crowbar/template-keystone.json``. Add the new
+parameter there as well. Here you should also increment the schema revision
+number.
+
+Add a schema migration in ``/opt/dell/chef/data_bags/migrate/keystone/``. A
+``*_noop.rb`` migration can be copied for the function prototypes. The file must
+start with the new schema revision number that you incremented in the default
+data bag. The migration needs to take the new parameter from the template
+attributes hash and add it to the attributes hash. Use another migration that
+adds a parameter as an example.
```

Comment:
Could you use full name (it's an example after all) together with specific migration number?

Reply:
Can you give me an example of how you would like it to change? Are you looking for:

> Add a schema migration in ``/opt/dell/chef/data_bags/migrate/keystone/302_add_insecure_debug.rb``. The ``300_noop.rb`` migration can be copied...

My only concern is that the number will quickly become out of date.

Reply:
Tried to add more specifics, let me know what you think.

Reply:
I like it. Yes, the number will become out of date, but I think it's better this way.